File: 198. House Robber.java
public class Solution {
    public int rob(int[] nums) {
        int output = 0;
        Map<Integer, Integer> hm = new HashMap<Integer, Integer>();
        if(nums.length == 0) {
            return 0;
        }
        output = house(nums, 0, hm);
        
        return output;
    }
    
    public int house(int[] nums, int index, Map<Integer, Integer> hm) {
        int sum = 0;
        if (index >= nums.length) {
            return sum;
        }
        
        if (hm.containsKey(index)) {
            return hm.get(index);
        } else {
            sum = nums[index];
            int l = sum + house(nums, index + 2, hm);
            int r = house(nums, index + 1, hm);
            sum = Math.max(l,r); 
            hm.put(index, sum);
        }
        
        return sum;
    }
}

File: 198. House Robber.py
class Solution:
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        def canRob(curr, nums, memo):
            if curr >= len(nums):
                return 0
            if curr in memo:
                return memo[curr]
            take = nums[curr] + canRob(curr + 2, nums, memo)
            skip = canRob(curr + 1, nums, memo)
            memo[curr] = max(take, skip)
            
            return memo[curr]
        
        return canRob(0, nums, {})
